Mathlib/Algebra/Group/UniqueProds.lean

```diff
@@ -196,36 +216,153 @@ attribute [to_additive UniqueSums] UniqueProds
 namespace Multiplicative
 
 instance {M} [Add M] [UniqueSums M] : UniqueProds (Multiplicative M) where
-  uniqueMul_of_nonempty {A} {B} hA hB := by
-    let A' : Finset M := A
-    have hA' : A'.Nonempty := hA
-    obtain ⟨a0, hA0, b0, hB0, J⟩ := UniqueSums.uniqueAdd_of_nonempty hA' hB
-    exact ⟨ofAdd a0, hA0, ofAdd b0, hB0, fun a b aA bB H ↦ J aA bB H⟩
+  uniqueMul_of_nonempty := UniqueSums.uniqueAdd_of_nonempty (G := M)
 
 end Multiplicative
 
 namespace Additive
 
 instance {M} [Mul M] [UniqueProds M] : UniqueSums (Additive M) where
-  uniqueAdd_of_nonempty {A} {B} hA hB := by
-    let A' : Finset M := A
-    have hA' : A'.Nonempty := hA
-    obtain ⟨a0, hA0, b0, hB0, J⟩ := UniqueProds.uniqueMul_of_nonempty hA' hB
-    exact ⟨ofMul a0, hA0, ofMul b0, hB0, fun a b aA bB H ↦ J aA bB H⟩
+  uniqueAdd_of_nonempty := UniqueProds.uniqueMul_of_nonempty (G := M)
 
 end Additive
 
+#noalign covariants.to_unique_prods
+#noalign covariants.to_unique_sums
+
+namespace UniqueProds
+
+open Finset MulOpposite in
+@[to_additive]
+theorem of_mulOpposite (G : Type*) [Mul G] (h : @UniqueProds Gᵐᵒᵖ (MulOpposite.mul G)) :
+    UniqueProds G :=
+⟨fun hA hB =>
+  let f : G ↪ Gᵐᵒᵖ := ⟨op, op_injective⟩
+  let ⟨y, yB, x, xA, hxy⟩ := h.uniqueMul_of_nonempty (hB.map (f := f)) (hA.map (f := f))
+  ⟨unop x, (mem_map' _).mp xA, unop y, (mem_map' _).mp yB, hxy.of_mulOpposite⟩⟩
+
 -- see Note [lower instance priority]
-/-- This instance asserts that if `A` has a multiplication, a linear order, and multiplication
-is 'very monotone', then `A` also has `UniqueProds`. -/
-@[to_additive Covariants.to_uniqueSums
-      "This instance asserts that if `A` has an addition, a linear order, and addition
-is 'very monotone', then `A` also has `UniqueSums`."]
-instance (priority := 100) Covariants.to_uniqueProds {A} [Mul A] [LinearOrder A]
-    [CovariantClass A A (· * ·) (· ≤ ·)] [CovariantClass A A (Function.swap (· * ·)) (· < ·)]
-    [ContravariantClass A A (· * ·) (· ≤ ·)] : UniqueProds A where
-      uniqueMul_of_nonempty {A} {B} hA hB :=
-        ⟨_, A.min'_mem ‹_›, _, B.min'_mem ‹_›, fun a b ha hb ab ↦
-        eq_and_eq_of_le_of_le_of_mul_le (Finset.min'_le _ _ ‹_›) (Finset.min'_le _ _ ‹_›) ab.le⟩
-#align covariants.to_unique_prods Covariants.to_uniqueProds
-#align covariants.to_unique_sums Covariants.to_uniqueSums
+/-- This instance asserts that if `A` has a right-cancellative multiplication, a linear order,
+  and multiplication is strictly monotone w.r.t. the second argument, then `A` has `UniqueProds`. -/
+@[to_additive _root_.Covariant.to_uniqueSums_right
+  "This instance asserts that if `A` has a right-cancellative addition, a linear order,
+  and addition is strictly monotone w.r.t. the second argument, then `A` has `UniqueSums`." ]
+instance (priority := 100) _root_.Covariant.to_uniqueProds_right {A} [Mul A] [IsRightCancelMul A]
```

Comment:
Maybe this could be `UniqueProds.of_Covariant_right` because we're in `UniqueProds` namespace.

Reply:
Ok, I am not terribly fussed by the name of the instance, since it should mostly be invisible, possibly will be called out explicitly a couple of times.  Iĺl make the change.